### PR TITLE
refactor: remove unstable Rust features

### LINKED ISSUE #130: Fails to build with Rust stable
Hello! 👋🏼 

I [packaged](https://archlinux.org/packages/extra/x86_64/hawkeye/) this tool for Arch Linux, however I'm hitting:

```
error[E0554]: `#![feature]` may not be used on the stable release channel
  --> fmt/src/lib.rs:15:1
   |
15 | #![feature(extract_if)]
   | ^^^^^^^^^^^^^^^^^^^^^^^

error[E0554]: `#![feature]` may not be used on the stable release channel
  --> fmt/src/lib.rs:16:1
   |
16 | #![feature(let_chains)]
   | ^^^^^^^^^^^^^^^^^^^^^^^

error[E0554]: `#![feature]` may not be used on the stable release channel
  --> fmt/src/lib.rs:15:12
   |
15 | #![feature(extract_if)]
   |            ^^^^^^^^^^
```

I solved this by using `cargo-nightly` package but this unfortunately effects reproducibility :/

I was wondering if it is possible to support Rust stable 👀

### PR DESCRIPTION
**Pull request by orhun**
closes #130

### CODE REVIEW

File: fmt/src/header/parser.rs [hunk 1]
```diff
@@ -43,18 +43,20 @@ pub fn parse_header(
         let mut begin_pos = 0;
         if header_def.skip_line_pattern.is_some() {
             // the format expect to find lines to be skipped
-            while let Some(l) = line.as_ref()
-                && !header_def.is_skip_line(l)
-            {
+            while let Some(l) = line.as_ref() {
+                if header_def.is_skip_line(l) {
```
Comment (tisonkun) [inline fmt/src/header/parser.rs:47]: Seems boolean logic reverted.

Please try to fix by making CI passed.

File: fmt/src/header/parser.rs [hunk 2]
```diff
@@ -71,29 +73,33 @@ pub fn parse_header(
     };
 
     // 2. has header
-    let existing_header = {
+    let mut existing_header = || -> bool {
```
Comment (tisonkun) [inline fmt/src/header/parser.rs:76]: Why this change needed? I tend to avoid "top" level mut.

This block is somewhat a function but I rush it without handling argument passing.

File: fmt/src/header/parser.rs [hunk 3]
```diff
@@ -71,29 +73,33 @@ pub fn parse_header(
...
         }
 
         // check if there is already a header
         let mut got_header = false;
-        if let Some(l) = line.as_ref()
-            && header_def.is_first_header_line(l)
-        {
+        if let Some(l) = line.as_ref() {
+            if header_def.is_first_header_line(l) {
+                return false;
```
Comment (tisonkun) [inline fmt/src/header/parser.rs:89]: OK. I see that return is helpful to ensure short circuit logic. You may try to extract this block into a function, rather than keep it a closure, if you would go this way.

File: fmt/src/selection.rs [hunk 4]
```diff
@@ -75,13 +75,14 @@ impl Selection {
...
         let (excludes, reverse_excludes) = {
             let mut excludes = self.excludes;
-            let reverse_excludes = excludes
-                .extract_if(|pat| pat.starts_with('!'))
-                .map(|mut pat| {
-                    pat.remove(0);
-                    pat
-                })
-                .collect::<Vec<_>>();
+            let mut reverse_excludes = Vec::new();
```
Comment (tisonkun) [inline fmt/src/selection.rs:78]: I tend to reimpl drain_filter as a function and add a to-do that we can replace it once the unstable feature becomes stable.

File: rust-toolchain.toml [hunk 5]
```diff
@@ -13,5 +13,5 @@
 # limitations under the License.
 
 [toolchain]
-channel = "nightly-2024-03-05"
+channel = "stable"
```
Comment (tisonkun) [inline rust-toolchain.toml:16]: Thank you! This guards the result.

I'm not quite familiar with stable, so I wonder if "1.76.0" more clear than "stable" and we may  bump explicitly?

Comment (alerque) [reply rust-toolchain.toml:16]: Actually no, just *stable* is the right answer here. One stabilized language features are not removed and you don't want to lock it to an older toolchain than the current stable one.

What you are after might be the MSRV value in Cargo.toml `[package.rust-version]`. That you can use to specify an exact minimum version.

Comment (tisonkun) [reply rust-toolchain.toml:16]: Thanks for your information @alerque!

File: fmt/src/header/parser.rs (hunk 1 above)
Comment (tisonkun) [reply fmt/src/header/parser.rs:47]: Emm ... Not inverted here. Boolean logic is too hard to me :<

File: fmt/src/header/parser.rs (hunk 2 above)
Comment (orhun) [reply fmt/src/header/parser.rs:76]: I added this for the sake of keeping to code clean. Otherwise I needed to add a new `if` block which will push the indentation further 💀 Now I can early return from this closure.

But I can remove it and simply use and `if` if you prefer that.

File: fmt/src/selection.rs (hunk 4 above)
Comment (orhun) [reply fmt/src/selection.rs:78]: Can you elaborate? `drain_filter` is an unstable feature, no? Do you mean wrapping this functionality in another function?

**Comment by orhun** [comment]
Any idea why the CI fails? Is there a wrong boolean logic somewhere?

File: fmt/src/header/parser.rs [hunk 6]
```diff
@@ -120,13 +126,15 @@ pub fn parse_header(
...
                             line = file_content.next_line();
-                            if let Some(l) = line.as_ref()
-                                && l.starts_with(before)
-                            {
-                                in_place_header.push_str(&l.to_lowercase());
-                                if header_def.multiple_lines && header_def.is_last_header_line(l) {
-                                    found_end = true;
-                                    break;
+                            if let Some(l) = line.as_ref() {
+                                if l.starts_with(before) {
```
Comment (tisonkun) [inline fmt/src/header/parser.rs:130]: This is the boolean logic error. The condition of else below is changed.

I'm patching this PR now, so you can wait for my follow-up.

File: rust-toolchain.toml (hunk 5 above)
Comment (tisonkun) [reply rust-toolchain.toml:16]: Added. `rust-version`.

Thanks again for your sharing :D

File: fmt/src/selection.rs (hunk 4 above)
Comment (tisonkun) [reply fmt/src/selection.rs]: I found it's not easy and not worthy to add a generic function, so I just add a todo on the block - https://github.com/korandoru/hawkeye/pull/132/commits/c574d1081b939069267268b36c446199692a6442

File: fmt/src/header/parser.rs (hunk 2 above)
Comment (tisonkun) [reply fmt/src/header/parser.rs]: I factor out this block into a function then.

**Comment by tisonkun** [comment]
All seems well. Going to merge.

@orhun Thanks for your contribution! Altough you may find I revert your commit, it's for a clean base to redo your tasks with my revisions. This patch is still heavily based on your work.
